### ISSUE TITLE
Shapes `to_labels` doesn't take into account edge width

### ISSUE THREAD

**Issue by psobolewskiPhD** [napari/napari#7652]
### 🐛 Bug Report

Reported https://forum.image.sc/t/binary-mask-from-lines-in-shapes-layer/109212/2

If you draw lines/paths and convert to Labels, the edge width is not considered.

### 💡 Steps to Reproduce

1. run napari
2. add a Shapes layer
3. draw a few lines or paths, adjusting the edge width
4. use layer list contextual menu: Convert to Labels and notice the edge width is disregarded

### 💡 Expected Behavior

The labels should reflect edge width.

### 🌎 Environment

napari: 0.5.6rc1.dev13+ga6d3eb52e.d20250202
Platform: macOS-14.5-arm64-arm-64bit
System: MacOS 14.5
Python: 3.12.8 | packaged by conda-forge | (main, Dec 5 2024, 14:19:53) [Clang 18.1.8 ]
Qt: 6.8.1
PyQt6: 
NumPy: 2.0.2
SciPy: 1.15.0
Dask: 2024.12.1
VisPy: 0.14.3
magicgui: 0.10.0
superqt: 0.7.2.dev4+g7850e53
in-n-out: 0.2.1
app-model: 0.3.1
psygnal: 0.11.1
npe2: 0.7.7
pydantic: 2.10.5

OpenGL:
- GL version: 2.1 Metal - 88.1
- MAX_TEXTURE_SIZE: 16384
- GL_MAX_3D_TEXTURE_SIZE: 2048

Screens:
- screen 1: resolution 1680x1050, scale 2.0

Optional:
- numba: 0.60.0
- triangle: 20250106
- napari-plugin-manager: 0.1.4

Settings path:
- /Users/piotrsobolewski/Library/Application Support/napari/napari-dev_a1eb8b76ba95fa16ad06e26097b46b8455dfbf0b/settings.yaml

**Comment by psobolewskiPhD** [comment]
Probably need to account for edge width here:
https://github.com/napari/napari/blob/39b385d23e9f0af40dcfea205d7e3a5d52feedb8/napari/layers/shapes/_shapes_utils.py#L791

I think for polygons and other filled shapes, edge width is less important, but for lines/paths, it's pretty crucial.

**Comment by brisvag** [comment]
I would actually disagree on this. The edge is just a visualisation aid, and its width does not reflect the actual points that the user has clicked, but instead includes an extra offset.

Even worse, if shapes were created numerically and programmatically, it would be surprising to suddenly have the shapes be bigger than expected just because the edge is there to help visualize.

**Comment by brisvag** [comment]
As for the forum post: we should probably do "something" for paths (maybe even just warn that they will be discarded), but thickening everything else is a bad side effect IMO.

**Comment by Czaki** [comment]
I need to support Lorenzo. 

The edges are to improve readability depending on zoom level and relative scale. If someone needs to get a thicker line, then the person should use dilation https://scikit-image.org/docs/stable/api/skimage.morphology.html#skimage.morphology.dilation

**Comment by psobolewskiPhD** [comment]
@Czaki Obviously dilation can work! By that same token, no-one needs Shapes to Labels, you can just DIY programmatically! But Shapes to Labels is a very handy GUI convenience function!

@brisvag What do you mean "polygons are specifically filled (that's data)"? there is nothing about polgygon data that implies a fill. Drawing them as a closed path is the simplest interpretation of the the data, which just defines vertexes. Filling them when rasterizing is a napari decision. It's arbitrary, but of course most people would probably expect it, given the context of a labels layer being typically used for masks. Arguably, this should be controllable, at which point polygons would be represented by vertexes and edges only and we *would* want edge width to work!

Returning to edges, the current rasterization uses 1 px for edge width. IMO this is arbitrary and letting the user control it via the edge width slider totally makes sense. It wouldn't change any existing programmatic behaviors. And if the end user didn't like how edge width = 10 looked after rasterization they could delete the layer and go again: I'm pretty sure it would be easy to leap to the conclusion that the phat line was due to it being phat on the Shapes layer.

I don't mean to be argumentative, so my apologies if I'm being overly pedantic here, but I do think there is value in consistency and thinking through behaviors -- particularly GUI ones which tend to be less flexible and used by less advanced users.
And, while of this sounds like minutia, we have bounced around the idea of unifying annotation, with labels being "painted" as vectors and then rasterized, so these behaviors are very relevant in that context.

**Comment by Czaki** [comment]
> [@brisvag](https://github.com/brisvag) What do you mean "polygons are specifically filled (that's data)"? there is nothing about polgygon data that implies a fill. Drawing them as a closed path is the simplest interpretation of the the data, which just defines vertexes. Filling them when rasterizing is a napari decision. It's arbitrary, but of course most people would probably expect it, given the context of a labels layer being typically used for masks. Arguably, this should be controllable, at which point polygons could be represented by vertexes and edges only and we _would_ want edge width to work!

It is decided by creator of data. If he selects data to be `polygon`, not `path`, then it contains all inner points by polygon definition. https://en.wikipedia.org/wiki/Polygon

**Comment by psobolewskiPhD** [comment]
You got me! 
But that's not really how our Shapes layer behaves?
I mean our path tool doesn't provide a way to make a closed shape, not that I can figure out anyways.
So the only way to make a closed shape is to use the rect, ellipse, and polygon tools.
And in these there is a clear distinction between the vertexes and path and the fill.

Anyhow, we're getting into the weeds here a bit -- the more I think about it, it should be possible to rasterize to draw an unfilled rect/ellipse/polygon. 😄

**Comment by brisvag** [comment]
> @brisvag What do you mean "polygons are specifically filled (that's data)"? 

Yeah I meant what @Czaki said :) It is controllable, because users can just choose `Path` instead.

> Returning to edges, the current rasterization uses 1 px for edge width. IMO this is arbitrary and letting the user control it via the edge width slider totally makes sense.

Ouch, didn't realize that it uses 1 px always... agree that this should either be 0, or equal to the edge width.

To clarify why I'm pusing against: I think the most common case scenario is this:

- user makes polygons to select areas, zooming in to get a better look and making sure the vertices are exactly where they want
- zoom out, check around, change edge width arbitrarily to make sure everything is right
- export to mask with some smallish edge width like 3-4
- never notice that there is a difference
- do calculations on now wrong masks

I find this much more likely than people thinking that the edge width is actually meant to be included, considering especially that users to generate masks are probably trying to do something *quantitative*.

> the more I think about it, it should be possible to rasterize to draw an unfilled rect/ellipse/polygon. 

I'm not sure what cases might want this, but even if it is useful to some, I think it should be a separate thing (maybe even another method like "mask from edges".

**Comment by jni** [comment]
Sorry I'm 100% with Lorenzo here. The data are the shapes — polygon, line, ellipse, etc — and edge width, color, etc, are visualisation parameters. In the grammar of graphics idioms, the polygon is data, and the line width is an aesthetic ([ref](https://rdrr.io/github/tidyverse/ggplot2/man/geom_polygon.html)). In GeoJSON, you don't store line width in the data, though you can store it as a "style" ([ref](https://gis.stackexchange.com/questions/22474/geojson-styling-information)).

Paths are a bit of a special case, in that when you rasterise a 0D object you *necessarily* have to give it a thickness that is defined by the resolution of your rasterisation. But that does not imply that we should add the thickness to the rasterisation, which would then be inconsistent with the polygons.

> the more I think about it, it should be possible to rasterize to draw an unfilled rect/ellipse/polygon.

these should then be a different type of object (which would not have face rendering regardless of face color settings, because it wouldn't have a face).

**Comment by psobolewskiPhD** [comment]
OK, I still think it's arbitrary that we provide a good default for rasterizing drawn shapes (polygons, ellipse) but not lines.
I will also add that I don't think most users have any idea of these "pure" concepts -- clearly I don't for that matter.
They draw a square, click Shapes to Labels and are happy. They draw a Square and a line and click the same button and are confused.
Anyhow, the workaround is to just use a polygon for line annotations, this seems a bit silly, but it works and lets the user control thickness.

**Comment by DragaDoncila** [comment]
> I will also add that I don't think most users have any idea of these "pure" concepts -- clearly I don't for that matter.
They draw a square, click Shapes to Labels and are happy

I'm kind of with Peter on this... I understand the theory underpinning the "0-width" polygon edges, but I think a user would be surprised for a big fat edge width to result in a thin polygon when rasterized. Furthermore I don't think their immediate thought would be "ah but of course, the polygon is an illusion, let me switch it to a path". So I do think for a GUI user we should consider how we can expose that aspect, if we do decide not to raster the edge widths for polygons.

**Comment by Czaki** [comment]
We may introduce additional action (convert to labels reflecting edges width) 

My expectation for converting path to labels is to have single dot width.

**Comment by TimMonko** [comment]
So with the analogy from Juan about pure vs styled. My, Draga's, and Peter's notion that GUI users would expect a stylized polygon. And Grzegorz suggestion about introducing an additional action we could...
1. Keep `to_labels` and its present 'pure' functionality. 
2. Add `style_to_labels` (or similar) to add the styled functionality. So, edge_width would be respected during the conversion. This would of course be targeted at using for lines, but someone could also do this for any polygon. 

> Anyhow, we're getting into the weeds here a bit -- the more I think about it, it should be possible to rasterize to draw an unfilled rect/ellipse/polygon.

We _could_ also add this functionality to `style_to_labels` whereby passing None (or some equivalent, like 0 alpha) to face_color, it would result in a polygon with only the styled edges drawn.

**Comment by psobolewskiPhD** [comment]
I really didn't want to get back to this, but:
*If* Shapes to Labels menu item is used, nothing destructive happens! The Shapes layer is preserved--this was changed a little while ago. 

Adding a new layer action seems like needless cognitive load to me. 
If Shapes to Layers is updated to use edge for paths, anyone using default edges will see no difference. 
Anyone using non-defaults will see their effect and can adjust that as needed based on their expectations.

**Comment by psobolewskiPhD** [comment]
> What I don't understand is: are we saying that people are actually clicking slighlty offset compared to where they want to click in order to account for the thickness of the visualized edges?

Not sure I follow this one. The image.sc post is linked above. The user at JAX that has run into wants straight lines, which is hard with labels layer. But when using shapes to labels they are 1 px, which on her images is invisible.
The workaround she is using is the rectangle tool, but I think we can all agree this is clumsy and if you want segmented lines you need the polygon tool.

**Comment by jni** [comment]
@psobolewskiPhD 

> I will also add that I don't think most users have any idea of these "pure" concepts -- clearly I don't for that matter.

I'm actually reasonably ready to die on this hill. 😂 I agree and understand that many people might not know them, but the thing with "pure" concepts is that they are teachable and ultimately make sense. This means that once you understand things, you can predict what's going to happen before you do it, even for new combinations. So with:

> They draw a square, click Shapes to Labels and are happy. They draw a Square and a line and click the same button and are confused.

I had to go and try it because I thought you were saying that the line wouldn't be drawn. But it *is* drawn, it just has no thickness. Which happened to be what I expected.

A good example for me is the JAX type promotion documentation, which honestly I consider a work of art:

https://docs.jax.dev/en/latest/jep/9407-type-promotion.html

In NumPy, you can't really predict what will happen when you add arrays of different dtypes, because everything is special cased. (The situation does seem to be improved in np2 since the Jax doc was written.) In JAX, there is a theory behind the code, so you *could* predict things before you try them out, if you understand the theory. (tbh, I did have to refresh my memory. 😂)

napari has grown to be quite a hodgepodge. But we *should* strive to be theoretically consistent with visualisation theory, so that eventually we can have folks grok a small number of things and from that be able to use napari consistently.

**Comment by psobolewskiPhD** [comment]
> I had to go and try it because I thought you were saying that the line wouldn't be drawn. But it is drawn, it just has no thickness. Which happened to be what I expected.

What you expected but not what users expect when they have adjusted the line to be thicker!

Like this isn't some hypothetical or theoretical, it's been reported on image.sc and I've had it crop up when supporting a user. Go figure, drawing straight lines is important.
So if we want to keep shape purity then we should add a path and line tool to the Labels layer.
Then we're duplicating all the tools between layers. Personally I think having rasterization work better from shape->label would enable us to simplify.

**Comment by Czaki** [comment]
@psobolewskiPhD Could you prepare use case story. So what user wants to achieve, rather than how he would like to achieve this?  



> Then we're duplicating all the tools between layers. Personally I think having rasterization work better from shape->label would enable us to simplify.

This is on our roadmap, to allow share painting tools between shapes and labels.

**Comment by psobolewskiPhD** [comment]
I'll ask for more details next time I run into the user.
The gist is they are annotating landmarks for an existing, downstream pipeline -- ML training I believe.
They need a raster input, so they tried Labels layer, but couldn't figure out how to draw straight lines. I suggested using Shapes and then convert to Labels and we ran into this issue.

**Comment by psobolewskiPhD** [comment]
Sure, but in the mean time if Shapes to Labels contextual rasterized lines as expected, everything would be gucci.
